### PR TITLE
swift-api-extract to generate JSON API information

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Add a new swift-frontend driver option that extract APIs in the swift
module and print in JSON format. This is to allow tooling to understand
and process swift APIs without the need to be a swift compiler or
understand swift module/AST.

Create a swift driver option that based on TBDGen but provide more information about swift APIs in both swift/objc/etc. in JSON format. The output can be easily consumed by other tools and scripts to reason about APIs in the swift module.

rdar://69600000

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci Please smoke test

File: lib/TBDGen/APIGen.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+namespace llvm {
+class raw_ostream;
+}
+
+namespace swift {
+namespace apigen {
+LLVM_ENABLE_BITMASK_ENUMS_IN_NAMESPACE();
+enum class APIAccess : uint8_t {
+  Unknown = 0, // Unknown, example, from binary.
+  Project = 1, // APIs available in project headers.
```
Comment (cyndyishida) [inline lib/TBDGen/APIGen.h:32]: This won't actually be looking at headers, we should update the comments.

**Comment by cachemeifyoucan** [comment]
ping

**Review by cyndyishida — APPROVED** [review]
LGTM,
I know theres some functionality/API's being used that already exist in `tapi`, do you see in the future us refactoring this out once `tapi` is fully merged into llvm? 
Regardless of the case, I'm wondering if theres value in documenting what details aren't currently applicable in a swift context, e.g. Project level APIAccess.

**Review by ributzka — APPROVED** [review]
LGTM

**Comment by cachemeifyoucan** [comment]
Rebase the change and run tests.

@nkcsgexi @DougGregor is this good enough to merge?

**Comment by cachemeifyoucan** [comment]
@swift-ci Please smoke test

**Review by nkcsgexi — APPROVED** [review]
LGTM!

**Comment by cachemeifyoucan** [comment]
@swift-ci Please smoke test

**Comment by akyrtzi** [comment]
Hi, sorry for the late question but have you considered using the JSON output of `swift-symbolgraph-extract`?

I'd like to see if it's possible to converge on a single tool for producing JSON output of Swift module APIs, instead of maintaining multiple incompatible tools.

**Comment by cachemeifyoucan** [comment]
> Hi, sorry for the late question but have you considered using the JSON output of `swift-symbolgraph-extract`?
> 
> I'd like to see if it's possible to converge on a single tool for producing JSON output of Swift module APIs, instead of maintaining multiple incompatible tools.

Yes, we have considered that. Our original methods includes writing a tool that converts symbolgraph JSON to API JSON and we eventually decide to do it this way. Symbolgraph does missing some information to compute all the information we need but we also want to push a bit further with api-extract tool to detach that from swift AST so it can be language agnostics to add stuffs that is more than swift. In the future, we might converge those two because we are going to add more fields to API JSON to include all the information needed by documentation from symbolgraph.

**Comment by migueldeicaza** [comment]
I found a crash when using the tool, filed it here:

https://bugs.swift.org/browse/SR-14311?jql=text%20~%20%22swift-api-extract%22

**Comment by cachemeifyoucan** [comment]
> I found a crash when using the tool, filed it here:
> 
> https://bugs.swift.org/browse/SR-14311?jql=text%20~%20%22swift-api-extract%22

Thanks for the report. I comment on the issue.

**Pull request merged**